cdap-app-templates/cdap-etl/cdap-data-pipeline/src/test/java/co/cask/cdap/datapipeline/DataPipelineTest.java

```diff
@@ -312,17 +322,110 @@ public void testExternalSparkProgramPipelines() throws Exception {
+    StructuredRecord recordSamuel = StructuredRecord.builder(schema).set("name", "samuel").build();
+    StructuredRecord recordKey1Value = StructuredRecord.builder(schema).set("name", "macroValue").build();
+    StructuredRecord recordKey2Value = StructuredRecord.builder(schema).set("name", "action1.row").build();
+
+    // write one record to each source
+    DataSetManager<Table> inputManager = getDataset(sourceName);
+    MockSource.writeInput(inputManager, ImmutableList.of(recordSamuel, recordKey1Value, recordKey2Value));
+
+    String defaultNamespace = NamespaceId.DEFAULT.getNamespace();
+    String triggeringPipeline = "macroActionTest-SPARK";
+    // Use properties from the triggering pipeline as values for runtime argument key1, key2, and key3
+    TriggeringPipelinePropertyMapping propertyMapping = new TriggeringPipelinePropertyMapping();
+
+      // use the value of runtime argument with key "value" in triggering pipeline as the value for key1
+    propertyMapping.addArgumentMapping(new TriggeringPipelinePropertyMapping.ArgumentMapping("value", key1));
+      // use the value of property "rowKey" in plugin "action1" in triggering pipeline as the value for key2
+    propertyMapping.addPluginPropertyMapping(
+      new TriggeringPipelinePropertyMapping.PluginPropertyMapping("action1", "rowKey", key2));
+    ProgramStatusTrigger completeTrigger =
+      new ProgramStatusTrigger(new WorkflowId(defaultNamespace, triggeringPipeline, SmartWorkflow.NAME),
+                               ImmutableSet.of(ProgramStatus.COMPLETED));
+    ScheduleId scheduleId = appId.schedule(("completeSchedule"));
+    appManager.addSchedule(
+      new ScheduleDetail(scheduleId.getNamespace(), scheduleId.getApplication(), scheduleId.getVersion(),
+                         scheduleId.getSchedule(), "",
+                         new ScheduleProgramInfo(SchedulableProgramType.WORKFLOW, SmartWorkflow.NAME),
+                         ImmutableMap.of(SmartWorkflow.TRIGGERING_PROPERTIES_MAPPING, GSON.toJson(propertyMapping)),
+                         completeTrigger, ImmutableList.<Constraint>of(), Schedulers.JOB_QUEUE_TIMEOUT_MILLIS, null));
+    appManager.enableSchedule(scheduleId);
+    WorkflowManager manager = appManager.getWorkflowManager(SmartWorkflow.NAME);
+    return manager;
+  }
+
+  private void assertTriggeredPipelinesResult(WorkflowManager workflowManager, Engine engine) throws Exception {
+    workflowManager.waitForRun(ProgramRunStatus.COMPLETED, 3, TimeUnit.MINUTES);
+    List<RunRecord> runRecords = workflowManager.getHistory(ProgramRunStatus.COMPLETED);
+    Assert.assertEquals(1, runRecords.size());
+    String key1 = "trigger-runtime-arg";
+    String key2 = "trigger-plugin-property";
+    Map<String, List<WorkflowTokenDetail.NodeValueDetail>> tokenData =
+      workflowManager.getToken(runRecords.get(0).getPid(), null, null).getTokenData();
+    Assert.assertEquals("macroValue", tokenData.get(key1).get(0).getValue());
```

Comment:
the fact that it's in the workflow token is an implementation detail that you shouldn't be asserting. If we change the implementation, this test will fail. It is enough to check the data in the sink and data in the tables.